tools/generate-helpers/pg-table-bindings/store_test.go.tpl

```diff
@@ -164,14 +128,66 @@ func (s *{{$namePrefix}}StoreSuite) TestStore() {
         {{$name}}s = append({{.TrimmedType|lowerCamelCase}}s, {{.TrimmedType|lowerCamelCase}})
     }
 
-	{{- if (.Obj.IsDirectlyScoped)}}
-	s.ErrorIs(store.UpsertMany(withAccessToDifferentNsCtx, {{.TrimmedType|lowerCamelCase}}s), sac.ErrResourceAccessDenied)
-	{{- end }}
 	s.NoError(store.UpsertMany(ctx, {{.TrimmedType|lowerCamelCase}}s))
 
     {{.TrimmedType|lowerCamelCase}}Count, err = store.Count(ctx)
     s.NoError(err)
-    s.Equal({{.TrimmedType|lowerCamelCase}}Count, 200)
+    s.Equal(200, {{.TrimmedType|lowerCamelCase}}Count)
     {{- end }}
 }
 
+{{- if .Obj.IsDirectlyScoped }}
+func (s *{{$namePrefix}}StoreSuite) TestSAC() {
```

Comment:
Would it make sense to split the SAC tests with one test per exposed functionality (Upsert, Delete, Get, Exists, Count, GetIDs, ...) ?
This may mean the context generation would have to go to a dedicated function
```
func getSACContexts(parent context.Context, access storage.Access, obj *{{.Type}}) map[string]context.Context {
    return map[string]context.Context {
        "allAccess": sac.WithAllAccess(parent),
        "noAccess": sac.WithNoAccess(parent),
        "accessToDifferentNs": sac.WithGlobalAccessScopeChecker(parent,
            sac.AllowFixedScopes(
                sac.AccessModeScopeKeys(access),
                sac.ResourceScopeKeys(targetResource),
                sac.ClusterScopeKeys({{ "obj" | .Obj.GetClusterID }}),
                sac.NamespaceScopeKeys("unknown ns")
            )),
        // ...
    }
}
```
Each test could then create the target object, set it in the storage if needed and run tests on the target function with appropriate test values.
The generated contexts and test cases could also be adjusted for cluster scoped and namespace scoped resource types.
cluster -> full access, no access, full cluster access, partial cluster access, wrong cluster access
namespace -> full access, no access, full cluster access, wrong cluster access, namespace access, wrong namespace access

Reply:
I'm not sure. I'd rather do it when we add test for new functions since now I don't know how the test will look like and what could be extracted